### ISSUE TITLE
cast to int64

### ISSUE THREAD

**Issue by larevio** [mongodb/laravel-mongodb#2600]
- Laravel-mongodb Version: v4.0.0-ALPHA2
- PHP Version: 8.1
- laravel version : 10

### Description:
i have an external application using mongodb , and saving some unix timestamps in int64
i want to update some of that value from laravel but when i pass the update parameter it turns to int32 type
can i change it to int64 ?

`$q->update([
                '$set' => [
                    'permissions.$.expiry' =>  (Carbon::createFromTimestamp($q->first()->expiry)->add($time[1], $time[0])->timestamp)
                ]
            ])`

**Comment by noor1yasser9** [comment]
In MongoDB, the default BSON type for integers is `32-bit`. However, `MongoDB 4.2 `and later support `64-bit integer `(int64) types, which you can use for storing large integer values. To ensure that the value you're updating is stored as a `64-bit integer`, you can explicitly specify the type when updating the document.
Here's how you can update a field with a `64-bit integer` value using Laravel's Eloquent and the `laravel-mongodb` package.
```
use Carbon\Carbon;

$newExpiry = Carbon::createFromTimestamp($q->first()->expiry)->add($time[1], $time[0])->timestamp;

$q->update([
    '$set' => [
        'permissions.$.expiry' => [
            '$numberLong' => (string) $newExpiry, // Convert to string to ensure it's stored as int64
        ],
    ],
]);

```

**Comment by alcaeus** [comment]
@noor1yasser9 the solution you suggested is completely wrong and does not produce the expected result. What you have in mind is the extended JSON format, which is a way to serialise BSON to JSON while retaining type information. While an Int64 value does serialise to `{ "$numberLong": "123" }` in extended JSON, the PHP driver does not support using extended JSON in update mechanisms. So instead of setting the `expiry` field to an int value, you're actually setting it to an embedded document containing a `$numberLong` field with a string value.

TL;DR: If you want to ensure that integer values serialise as 64-bit integers in BSON regardless of their value, store them as `MongoDB\BSON\Int64` instances. Make sure to use `ext-mongodb` 1.16 or newer for this to work, as previous versions did not allow creating `Int64` instances. A detailed explanation follows.

---

The PHP driver encodes integer values as 32-bit values as long as they fall within the 32-bit range, and only uses an int64 BSON type if the value exceeds the 32-bit range. For most applications, this works fine, but if you're dealing with an application that always expects an int64 value, it can cause issues. With the 1.16 release of `ext-mongodb`, we've addressed this by allowing users to create instances of the `MongoDB\BSON\Int64` class, which was previously only used on 32-bit platforms to represent 64-bit values. You can now create instances of this class, and on serialisation the value will always be serialised as a 64-bit integer even if it falls within the 32-bit range.

Let me show this with an example. In the example, we insert a document, update it using the method suggested by @noor1yasser9, then read the document. The `['root' => 'bson']` typeMap option instructs the driver to return a `MongoDB\BSON\Document` instance instead of deserialising the BSON object to PHP. This allows us to inspect the actual types used in serialisation of the `x` field by calling `get('x')` on the document:
```php
$collection->insertOne(['_id' => 1]);
$collection->updateOne(['_id' => 1], ['$set' => ['x' => ['$numberLong' => '123']]]);

$document = $collection->findOne(['_id' => 1], ['typeMap' => ['root' => 'bson']]);
var_dump($document->get('x'));
```

This example prints the following:
```
class MongoDB\BSON\Document#722 (2) {
  public $data =>
  string(36) "GgAAAAIkbnVtYmVyTG9uZwAEAAAAMTIzAAA="
  public $value =>
  class stdClass#724 (1) {
    public $$numberLong =>
    string(3) "123"
  }
}
```

The `x` field contains an embedded document with a single `$numberLong` field as suggested in my first paragraph. You can see the actual document contents in the `value` property of the above output. The `data` property contains the binary BSON data as a base64 encoded string.

Let's change the `updateOne` call in the example above to use an `Int64` instance:
```php
$collection->updateOne(['_id' => 1], ['$set' => ['x' => new Int64('123')]]);
```

This yields the following value for `x`:
```
class MongoDB\BSON\Int64#722 (1) {
  public $integer =>
  string(3) "123"
}
```

The value is now correctly serialised as a 64-bit integer, despite it being in the 32-bit range. We can see this because when working with raw BSON, the driver retains the original types instead of converting to the appropriate PHP type (which in this case would be an `int`). Please be aware that when reading this document with a typeMap that deserialises BSON to PHP values (which is the default), the value will once again be returned as a PHP integer:
```
class stdClass#720 (2) {
  public $_id =>
  int(1)
  public $x =>
  int(123)
}
```

This is usually intended and would not cause issues; just make sure to always create an `Int64` instance when updating a value that you expect to always be an `Int64`.

**Comment by GromNaN** [comment]
I'm closing because there's been no response to the detailed explanation from @alcaeus. I consider the problem solved.
